### DISCUSSION TITLE
Windows Application Logs Question

### DISCUSSION THREAD

**Discussion by anthony-pepe** [wazuh/wazuh#15637] [Q&A] [answered]
We would like Wazuh to collect Windows Event logs > Application/CAPI2
Specifically like Windows Event ID 41, & 90
I was curious how do I have Wazuh collect those logs & write a rule or something so that it will ingest it & show up in the Agents > Events dashboard?

Thank you for your assistance

**Comment by jctello** [comment, marked as answer]
Hi @anthony-pepe ,
In order to collect these logs you may use the following configuration on the agent's `ossec.conf` file:
```
<ossec_config>
  <localfile>
    <location>Microsoft-Windows-CAPI2/Operational</location>
    <log_format>eventchannel</log_format>
  </localfile>
</ossec_config>
```
You may also distribute this configuration by using [centralized configuration](https://documentation.wazuh.com/current/user-manual/reference/centralized-configuration.html) 

With that all events in the CAPI2 Operational channel will be collected, you can also use [\<query\>](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/localfile.html#query) to filter which events get collected and reduce the usage of network traffic. 

In order for the events to be registered you indeed need to create rules so the Wazuh Analysis engine correctly identifies these events as being security relevant. You can do so by creating `/var/ossec/etc/rules/CAPI2_rules.xml` on the Wazuh manager with the following content:

```
<group name="CAPI2">

<rule id="100500" level="0">
<!-- Without custom rules the event matches the Windows informational event rule -->
  <if_sid>60009</if_sid>  
  <field name="win.system.channel">Microsoft-Windows-CAPI2/Operational</field>
  <description>Event from CAPI2 Diagnostics</description>
  <options>no_full_log</options>
</rule>

<rule id="100501" level="6">
  <if_sid>100500</if_sid>  
  <field name="win.system.eventID">41</field>
  <description>CAPI2: Revocation verification exception</description>
  <options>no_full_log</options>
</rule>

<rule id="100502" level="6">
  <if_sid>100500</if_sid>  
  <field name="win.system.eventID">90</field>
  <description>Active Directory Certificate Services detected an exception</description>
  <options>no_full_log</options>
</rule>

</group>
```

Thank you for your interesting question, let me know if this answered your query!

**Reply by anthony-pepe** [reply]
Thank you for that response, that is really helpful.

When wanting to use <query> what is the proper way to query only the EventID's listed above.
I referenced the <query> documentation page and its not 100% clear.

Additionally since implementing this rule above, we are now getting emails triggered from (Rule: 60014 fired (level 10) -> "Multiple Windows error events.")

What is the best way to mitigate this?
Thank you for your assistance

**Reply by jctello** [reply]
Hi @anthony-pepe , thank you for the feedback!

Given that there are different channels, and that certain Windows events are already picked up by previous rules it can be complex to determine which rule to base your custom rules on.

To create the rules above I enabled `<logall_json>` then used the `full_log` portion of the events observed on `archives.json` to test against the `/var/ossec/bin/wazuh-logtest-legacy` tool, but not without first translating the escaped quotes (`\"` ) and temporarily modifying rule `60000` to take in events decoded as `json` and removing the `<category>` criteria. We understand this process is less than ideal and a [new analysis engine is being created for Wazuh 5.0](https://github.com/wazuh/wazuh/issues/11334) which will include many improvements to this process.

As for the new events you are seeing, I believe they are caused due to rule `60014` which is triggered when rule `60011` is triggered 8 times in less than 4 minutes. Rule `60011` is looking for any windows event whose field `win.system.severityValue` has the value `ERROR`.

Mitigations depend on how you wish to approach this, creating a rule that catches all CAPI2 error events will ensure the composite rule is not considered:

```
<rule id="100503" level="6">
  <if_sid>100500</if_sid>  
  <field name="win.system.severityValue">^ERROR$</field>
  <description>CAPI2 error event</description>
  <options>no_full_log</options>
</rule>
```

Alternatively you may enhance the ruleset to identify the various types of errors observed, for example I note that on my test environment eventIDs 11,41,30 and 82 will trigger this frequently (specially eventID 11). So you may create additional rules such as:

```
<rule id="100504" level="6">
  <if_sid>100500</if_sid>  
  <field name="win.system.eventID">11</field>
  <description>Frequent CAPI2 certificate verification error</description>
  <options>no_full_log</options>
</rule>
```

Finally you may reduce the level of the rule being triggered so it does not match your email alert criteria:

```
  <rule id="60014" level="8" frequency="8" timeframe="240" overwrite="yes">
    <if_matched_sid>60011</if_matched_sid>
    <options>no_full_log</options>
   <options>no_email_alert</options>
    <description>Multiple Windows error events.</description>
  </rule>
```

**Reply by anthony-pepe** [reply]
When installing this code into ossec.conf on the agent - we are getting agent queue flooding emails towards the end of day.
What is the best way to filter (add query) to this code to only get rule 41, 90, 11, & 30?

`<localfile>
    <location>Microsoft-Windows-CAPI2/Operational</location>
    <log_format>eventchannel</log_format>
  </localfile>`
 

Does this look like what should be added?
`<query>Event/System[EventID = 41 or EventID = 11 or EventID = 90 or EventID = 30]</query>`

Thank you for your assistance

**Reply by jctello** [reply]
I've verified that query and that is absolutely correct. 
Full configuration would be:
```
<localfile>
  <location>Microsoft-Windows-CAPI2/Operational</location>
  <query>Event/System[EventID = 41 or EventID = 11 or EventID = 90 or EventID = 30]</query>
  <log_format>eventchannel</log_format>
</localfile>
```
Then you will only see events related to those specific IDs

**Reply by anthony-pepe** [reply]
Thank you for the clarification.

One final question, even with those event queryies we are still getting agent queue flooding. Any idea how to remediate this?
Received From: (X) any->wazuh-agent
Rule: 204 fired (level 12) -> "Agent event queue is flooded. Check the agent configuration."
Portion of the log(s):

wazuh: Agent buffer: 'flooded'.
level: flooded

**Reply by jctello** [reply]
Hi @anthony-pepe in this is the [anti-flooding mechanism](https://documentation.wazuh.com/current/user-manual/capabilities/antiflooding.html) at work. 
This may be caused by this particular channel or any other, it is worth investigating which is the dominant source of events by temporarily enabling `<logall_json>` in the manager and counting which are the most common messages observed.

If you determine that the events being sent by the agent are adequately filtered but you still need to frequently send more than 500 events per second from that source you may either increase [the client buffer settings](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/client-buffer.html) or disable this mechanism altogether.

**Reply by anthony-pepe** [reply]
Hello,

Yes, it's definitely this channel causing the flooding. I disabled the setting on one server & I no longer got the queue flooding emails.

I suppose that the only way to remediate this is to change the client buffer setting, by increasing the size or disabling it then?

**Reply by anthony-pepe** [reply]
@jctello Hello, I've increased the events per second to the maximum of 1000, I'm still getting the (Rule: 204 fired (level 12) -> "Agent event queue is flooded. Check the agent configuration.")

Any suggestions? I've checked the logs & there isn't anything from that CAPI2 events I can filter out.

Thanks in advance
